kolibri/content/utils/annotation.py

```diff
@@ -1,38 +1,163 @@
-
-        ch_metadata_obj, _ = ChannelMetadataCache.objects.update_or_create(
-            id=db_name,
-            defaults=update_values,
+    from .channel_import import import_channel_from_local_db
+    channel_ids = get_channel_ids_for_content_database_dir(settings.CONTENT_DATABASE_DIR)
+    for channel_id in channel_ids:
+        if not ChannelMetadata.objects.filter(id=channel_id).exists():
+            import_channel_from_local_db(channel_id)
+
+def set_leaf_node_availability_from_local_file_availability():
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
+
+    ContentNodeTable = bridge.get_table(ContentNode)
+    FileTable = bridge.get_table(File)
+    LocalFileTable = bridge.get_table(LocalFile)
+
+    connection = bridge.get_connection()
+
+    file_statement = select([LocalFileTable.c.available]).where(
+        and_(
+            FileTable.c.local_file_id == LocalFileTable.c.id,
+        )
+    ).limit(1)
+
+    logging.info('Setting availability of File objects based on LocalFile availability')
+
+    connection.execute(FileTable.update().values(available=file_statement).execution_options(autocommit=True))
+
+    contentnode_statement = select([FileTable.c.contentnode_id]).where(
+        and_(
+            FileTable.c.available == True,  # noqa
+            FileTable.c.supplementary == False
         )
+    )
+
+    logging.info('Setting availability of non-topic ContentNode objects based on File availability')
+
+    connection.execute(ContentNodeTable.update().where(
+        ContentNodeTable.c.id.in_(contentnode_statement)).values(available=True).execution_options(autocommit=True))
+
+    connection.execute(ContentNodeTable.update().where(
+        ContentNodeTable.c.id.notin_(contentnode_statement)).where(
+        ContentNodeTable.c.kind != content_kinds.TOPIC).values(available=False).execution_options(autocommit=True))
+
+    bridge.end()
+
+def set_local_file_availability(checksums):
+    """
+    Shortcut method to update database if we are sure that the files are available.
+    Can be used after successful downloads to flag availability without having to do expensive disk reads.
+    """
+    bridge = Bridge(app_name=CONTENT_APP_NAME)
+
+    LocalFileClass = bridge.get_class(LocalFile)
+
+    logging.info('Setting availability of {number} LocalFile objects based on passed in checksums'.format(number=len(checksums)))
 
-        # Records a new last_updated only if channel is brand new. Does not
-        # handle case where channel's version is upgraded, which is not
-        # yet supported on UI anyway
-        if ch_metadata_obj.last_updated is None:
-            ch_metadata_obj.last_updated = local_now()
-            ch_metadata_obj.save()
+    for i in range(1, int(math.ceil(len(checksums)/10000.0)) + 1):
+        bridge.session.bulk_update_mappings(LocalFileClass, ({
+            'id': checksum,
+            'available': True
+        } for checksum in checksums[0:i*10000]))
```

Comment:
Yes, that is an error...

I had done some testing with the KA English channel, and hadn't found any issues. Again, more empirical work needed!